core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java

```diff
@@ -0,0 +1,279 @@
+      return true;
+    }
+    return  false;
+  }
+
+
+  private void readAsync(final ByteBuffer byteBuffer) throws IOException {
+    stateChangeLock.lock();
+    if (endOfStream || isReadInProgress) {
+      stateChangeLock.unlock();
+      return;
+    }
+    byteBuffer.position(0);
+    byteBuffer.flip();
+    isReadInProgress = true;
+    stateChangeLock.unlock();
+    executorService.execute(() -> {
+      byte[] arr;
+      stateChangeLock.lock();
+      arr = byteBuffer.array();
+      stateChangeLock.unlock();
+      // Please note that it is safe to release the lock and read into the read ahead buffer
+      // because either of following two conditions will hold - 1. The active buffer has
+      // data available to read so the reader will not read from the read ahead buffer.
+      // 2. The active buffer is exhausted, in that case the reader waits for this async
+      // read to complete. So there is no race condition in both the situations.
+      int nRead = 0;
+      while (nRead == 0) {
+        try {
+          nRead = underlyingInputStream.read(arr);
+          if (nRead < 0) {
+            // We hit end of the underlying input stream
+            break;
+          }
+        } catch (Exception e) {
+          stateChangeLock.lock();
+          // We hit a read exception, which should be propagated to the reader
+          // in the next read() call.
+          isReadAborted = true;
+          readException = e;
+          stateChangeLock.unlock();
+        }
+      }
+      stateChangeLock.lock();
+      if (nRead < 0) {
+        endOfStream = true;
+      }
+      else {
+        // fill the byte buffer
+        byteBuffer.limit(nRead);
+      }
+      isReadInProgress = false;
+      signalAsyncReadComplete();
+      stateChangeLock.unlock();
+    });
+  }
+
+  private void signalAsyncReadComplete() {
+    stateChangeLock.lock();
+    try {
+      asyncReadComplete.signalAll();
+    } finally {
+      stateChangeLock.unlock();
+    }
+  }
+
+  private void waitForAsyncReadComplete() {
+    stateChangeLock.lock();
+    try {
+      asyncReadComplete.await();
+    } catch (InterruptedException e) {
+    } finally {
+      stateChangeLock.unlock();
+    }
+  }
+
+  @Override
+  public synchronized int read() throws IOException {
+    int val = read(oneByte, 0, 1);
+    if (val == -1) {
+      return -1;
+    }
+    return oneByte[0] & 0xFF;
+  }
+
+  @Override
+  public synchronized int read(byte[] b, int offset, int len) throws IOException {
+    stateChangeLock.lock();
+    try {
+      len = readInternal(b, offset, len);
+    }
+    finally {
+      stateChangeLock.unlock();
+    }
+    return len;
+  }
+
+  /**
+   * Internal read function which should be called only from read() api. The assumption is that
+   * the stateChangeLock is already acquired in the caller before calling this function.
+   */
+  private int readInternal(byte[] b, int offset, int len) throws IOException {
+
```

Comment:
Can we insert `assert(stateChangeLock.isLocked())` after declaring `stateChangeLock` as `ReentrantLock`?